File: python/blackjack.py
from os import readlink
import random

class Card:
  def __init__(self, rank, suit):
    self.rank = rank
    self.suit = suit


  def to_string(self):
    return f"{self.rank}{self.suit}"

  def get_points(self):
    if type(self.rank) == int:
      return int(self.rank)
    elif self.rank == 'A':
      return 11
    else:
      return 10



class Hand:
  def __init__(self, cards):
    self.cards = cards
    self.bust = False

  def get_points(self):
    if not self.cards:
      return 0
    elif all(card.rank == "A" for card in self.cards):
      return 21
    else:
      points = 0
      for card in self.cards:
        points += card.get_points()
      return points

  def to_string(self):
    return f"{', '.join([card.to_string() for card in self.cards])}"

  def hit(self, new_card):
    print("Hitting...")
    self.cards.append(new_card)

  def check_bust(self):
    if self.get_points() > 21:
      print("Bust.")
      self.bust = True
      return False
    else:
      return True



class Deck:
  def __init__(self):
    self.cards = self.generate_deck()

  def generate_deck(self):
    suits = ['S', 'D', 'C', 'H']
    ranks = ['A', 2, 3, 4, 5, 6, 7, 8, 9, 10, 'J', 'Q', 'K']
    deck = []
    for suit in suits:
      for rank in ranks: 
        deck.append(Card(rank, suit))
    return deck

  def shuffle(self):
    random.shuffle(self.cards)
  
  def draw(self):
    return self.cards.pop()


class Game:
  def __init__(self):
    self.deck = Deck()

  def get_winner(self, player_hand, dealer_hand):
    p_points = player_hand.get_points()
    d_points = dealer_hand.get_points()

    if (player_hand.bust and dealer_hand.bust) or (p_points == d_points):
      print("Draw!")
    elif (dealer_hand.bust) or (p_points > d_points):
      print("Player wins.")
    elif(player_hand.bust) or (p_points < d_points):
      print("Dealer wins.")


  def play(self):
    self.deck.shuffle()
    player_hand = Hand([self.deck.draw(), self.deck.draw()])
    print(f"Player hand: {player_hand.to_string()}. \n {player_hand.get_points()} points.")
    
    # Player turn
    player_turn = True
    while(player_turn):
      hit_or_stick = input("Enter 'hit' or 'stick': ")
      if hit_or_stick == "hit":
        player_hand.hit(self.deck.draw())
        print(f"Player hand: {player_hand.to_string()}. \n {player_hand.get_points()} points.")
        player_turn = player_hand.check_bust()
      else:
        player_turn = False

    # Dealer turn
    dealer_hand = Hand([self.deck.draw(), self.deck.draw()])
    print(f"Dealer hand: {dealer_hand.to_string()}. \n {dealer_hand.get_points()} points.")
    while(dealer_hand.get_points() <= 17):
      dealer_hand.hit(self.deck.draw())
      print(f"Dealer hand: {dealer_hand.to_string()}. \n {dealer_hand.get_points()} points.")
      dealer_hand.check_bust()

  
    # Get winner
    self.get_winner(player_hand, dealer_hand)






game = Game()
print(game.play())

File: typescript/blackjack.js
export class Card {
    constructor(rank, suit) {
        this.rank = rank;
        this.suit = suit.toUpperCase();
    }
    toString() {
        return `${this.rank}${this.suit}`;
    }
    get points() {
        if (parseInt(this.rank)) {
            return parseInt(this.rank);
        }
        else if (this.rank === 'A') {
            return 11;
        }
        else {
            return 10;
        }
    }
}
export class Deck {
    constructor() {
        this.cards = this.generateDeck();
    }
    generateDeck() {
        const suits = ['S', 'D', 'C', 'H'];
        const ranks = [
            'A',
            '2',
            '3',
            '4',
            '5',
            '6',
            '7',
            '8',
            '9',
            '10',
            'J',
            'Q',
            'K',
        ];
        const deck = [];
        for (let suit of suits) {
            for (let rank of ranks) {
                deck.push(new Card(rank, suit));
            }
        }
        return deck;
    }
    draw() {
        return this.cards.shift(); // Exclamation mark means can't be undefined
    }
    shuffle() {
        let seed = Math.floor(Math.random() * 10000 + 1);
        let currentIndex = this.cards.length;
        let temporaryValue;
        let randomIndex;
        let random = () => {
            var x = Math.sin(seed++) * 10000;
            return x - Math.floor(x);
        };
        while (0 !== currentIndex) {
            randomIndex = Math.floor(random() * currentIndex);
            currentIndex -= 1;
            temporaryValue = this.cards[currentIndex];
            this.cards[currentIndex] = this.cards[randomIndex];
            this.cards[randomIndex] = temporaryValue;
        }
        return this.cards;
    }
}
export class Hand {
    constructor(cards) {
        this.cards = cards;
        this.bust = false;
    }
    get points() {
        //Handle case for two Aces.
        if (this.cards.length === 0) {
            return 0;
        }
        else if (this.cards.every((card) => card.rank === 'A')) {
            return 21;
        }
        else {
            return this.cards.reduce((total, card) => {
                return (total += card.points);
            }, 0);
        }
    }
    toString() {
        return `${this.cards.join(', ')}\n(${this.points} points)`;
    }
    hit(newCard) {
        console.log('Hitting...');
        this.cards.push(newCard);
    }
    checkBust() {
        if (this.points > 21) {
            console.log('Bust.');
            this.bust = true;
            return false;
        }
        else {
            return true;
        }
    }
}
export class Game {
    constructor() {
        this.deck = new Deck();
    }
    play() {
        console.log('Welcome to a game of Blackjack.');
        this.deck.shuffle();
        const playerHand = this.playerTurn(new Hand([this.deck.draw(), this.deck.draw()]));
        const dealerHand = this.dealerTurn(new Hand([this.deck.draw(), this.deck.draw()]));
        // Get winner
        this.getWinner(playerHand, dealerHand);
    }
    playerTurn(hand) {
        console.log(`Your hand is ${hand.toString()}`);
        let playerTurn = true;
        while (playerTurn) {
            let hitOrStick = window.prompt("Enter 'hit' or 'stick': ");
            if (hitOrStick === 'hit') {
                hand.hit(this.deck.draw());
                console.log(`Your hand is ${hand.toString()}`);
                playerTurn = hand.checkBust();
            }
            else {
                playerTurn = false;
            }
        }
        return hand;
    }
    dealerTurn(hand) {
        console.log(`Dealers hand is ${hand.toString()}`);
        while (hand.points <= 17) {
            hand.hit(this.deck.draw());
            console.log(`Dealers hand is ${hand.toString()}`);
            hand.checkBust();
        }
        return hand;
    }
    getWinner(player, dealer) {
        const pPoints = player.points;
        const dPoints = dealer.points;
        if ((player.bust && dealer.bust) || pPoints === dPoints) {
            console.log('Draw!');
        }
        else if (dealer.bust || pPoints > dPoints) {
            console.log('You win!');
        }
        else if (player.bust || pPoints < dPoints) {
            console.log('Dealer wins.');
        }
    }
}
let game1 = new Game();
game1.play();
